File: app/src/main/java/com/ababilx/routinescrapper/data/StudentPrefs.kt
package com.ababilx.routinescrapper.data

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

private val Context.studentDataStore by preferencesDataStore(name = "student_prefs")

/** Remembers the last valid batch search so reopen feels instant. */
class StudentPrefs(private val context: Context) {
    val lastQuery: Flow<String> = context.studentDataStore.data.map { prefs ->
        prefs[KEY_LAST_QUERY].orEmpty()
    }

    suspend fun saveQuery(query: String) {
        val cleaned = query.trim().uppercase()
        context.studentDataStore.edit { prefs ->
            if (cleaned.isEmpty()) {
                prefs.remove(KEY_LAST_QUERY)
            } else {
                prefs[KEY_LAST_QUERY] = cleaned
            }
        }
    }

    companion object {
        private val KEY_LAST_QUERY = stringPreferencesKey("last_query")
    }
}

File: lib/data/student_prefs.dart
import 'package:shared_preferences/shared_preferences.dart';

class StudentPrefs {
  StudentPrefs(this._prefs);

  final SharedPreferences _prefs;

  static const _keyLastQuery = 'last_query';

  static Future<StudentPrefs> create() async {
    return StudentPrefs(await SharedPreferences.getInstance());
  }

  String lastQuery() => _prefs.getString(_keyLastQuery) ?? '';

  Future<void> saveQuery(String query) async {
    final cleaned = query.trim().toUpperCase();
    if (cleaned.isEmpty) {
      await _prefs.remove(_keyLastQuery);
    } else {
      await _prefs.setString(_keyLastQuery, cleaned);
    }
  }
}
